src/keys.rs

```diff
@@ -1,79 +1,110 @@
+                StatusCode::BAD_REQUEST,
+                "signature is not a valid hex string",
+            ),
+            Self::InvalidSignature => (
+                StatusCode::BAD_REQUEST,
+                "couldn't create signature from string",
+            ),
+        };
+        let body = Json(json!({
+            "error": error_message,
+        }));
+        (status, body).into_response()
+    }
 }
 
 pub struct Keys {
-    encoding: EncodingKey,
-    decoding: DecodingKey,
-    pubkey:   String,
+    wallet: LocalWallet,
 }
 
 pub type SharedKeys = Arc<Keys>;
 
 impl Keys {
     pub async fn new(options: &Options) -> Result<Self> {
-        info!(public_key = ?options.public_key, private_key=?options.private_key, "Loading JWT keys");
-        let (private_key, public_key) = try_join!(
-            tokio::fs::read(&options.private_key),
-            tokio::fs::read(&options.public_key)
-        )?;
-        Ok(Self {
-            encoding: EncodingKey::from_rsa_pem(&private_key)?,
-            decoding: DecodingKey::from_rsa_pem(&public_key)?,
-            pubkey:   String::from_utf8(public_key)?,
-        })
+        let phrase = options.mnemonic.as_ref();
+        let wallet = MnemonicBuilder::<English>::default()
+            .phrase(phrase)
+            .build()?;
+        Ok(Self { wallet })
     }
 
-    pub fn encode<T: Serialize>(&self, token: &T) -> Result<String, jsonwebtoken::errors::Error> {
-        encode(&Header::new(Self::alg()), token, &self.encoding)
+    pub async fn sign(&self, message: &str) -> Result<Signature, SignatureError> {
+        let signature = self
+            .wallet
+            .sign_message(message)
+            .await
+            .map_err(|_| SignatureError::SignatureCreation)?;
+        Ok(Signature(hex::encode::<Vec<u8>>(signature.into())))
     }
 
     #[allow(unused)]
-    pub fn decode<T: DeserializeOwned>(
-        &self,
-        token: &str,
-    ) -> Result<TokenData<T>, jsonwebtoken::errors::Error> {
-        decode::<T>(token, &self.decoding, &Validation::new(Self::alg()))
-    }
-
-    pub const fn alg_str() -> &'static str {
-        "PS256"
+    pub fn verify(&self, message: &str, signature: &Signature) -> Result<(), SignatureError> {
+        let h = hex::decode(&signature.0).map_err(|_| SignatureError::InvalidToken)?;
+        let signature = ethers_core::types::Signature::try_from(h.as_ref())
+            .map_err(|_| SignatureError::InvalidSignature)?;
+        signature
+            .verify(
+                RecoveryMessage::Data(message.as_bytes().to_owned()),
+                self.wallet.address(),
+            )
+            .map_err(|_| SignatureError::InvalidToken)
     }
 
-    fn alg() -> Algorithm {
-        Algorithm::from_str(Self::alg_str()).expect("unknown algorithm")
-    }
-
-    pub fn decode_key_to_string(&self) -> String {
-        self.pubkey.clone()
+    pub fn address(&self) -> String {
+        let adr = self.wallet.address();
+        hex::encode(adr)
```

Comment:
Ethereum addresses are not just hex encoded but require a specific pattern of capitalization of the digits.